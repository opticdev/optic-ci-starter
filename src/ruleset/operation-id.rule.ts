import {check} from "@useoptic/api-checks/build/sdk/define-check-test-dsl/define-check";
import {scenario} from "@useoptic/api-checks/build/sdk/define-check-test-dsl/scenarios";
import {expect} from "chai";

export default check('prevent operation id change')
  .description('changing an operation id is a breaking change')
  .implementation(({ operations }) => {
    const { expect } = require('chai');
    operations.changed.must('not change operation ids -- breaking', (before, after) => {
      if (before.operationId !== after.operationId) {
        expect.fail('expected operation id to be the same');
      }
    });
  })
  .failingExample(
    scenario('changing operation id').paths.changed(
      {
        '/example': {
          get: {
            operationId: '123',
            responses: {
              '200': {
                description: '',
              },
            },
          },
        },
      },
      (paths) => {
        paths['/example']!.get!.operationId = '456';
        return paths;
      }
    )
  );
