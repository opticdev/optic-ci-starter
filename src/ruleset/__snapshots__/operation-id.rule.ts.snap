// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`failing case 1: changing operation id 1`] = `
Object {
  "passed": false,
  "results": Array [
    Object {
      "change": Object {
        "changeType": "changed",
        "changed": Object {
          "after": Object {
            "method": "get",
            "operationId": "456",
            "pathPattern": "/example",
          },
          "before": Object {
            "method": "get",
            "operationId": "123",
            "pathPattern": "/example",
          },
        },
        "location": Object {
          "conceptualLocation": Object {
            "method": "get",
            "path": "/example",
          },
          "conceptualPath": Array [
            "operations",
            "/example",
            "get",
          ],
          "jsonPath": "/paths/~1example/get",
          "kind": "operation",
        },
      },
      "condition": "not change operation ids -- breaking",
      "docsLink": undefined,
      "effectiveOnDate": undefined,
      "error": "expected operation id to be the same",
      "isMust": true,
      "isShould": false,
      "passed": false,
      "where": "updated operation: GET /example",
    },
  ],
}
`;
